### PR TITLE
Cache intrinsic dimensions

### LINKED ISSUE #2298: IntrinsicWidth doesn't change its size when it's child's width changes.

### PR DESCRIPTION
**Pull request by Hixie**
Also, make sure that the parent is notified when they change.

Fixes #2298

### CODE REVIEW

**Comment by Hixie** [comment]
@abarth Given the issues I had with this patch, it might be worth you giving it a once-over again.

File: packages/flutter/lib/src/rendering/box.dart [hunk 1]
```diff
@@ -471,6 +471,28 @@ class BoxParentData extends ParentData {
...
+class _IntrinsicDimensionsCacheEntry {
+  _IntrinsicDimensionsCacheEntry(this.dimension, this.argument);
+
+  final _IntrinsicDimension dimension;
+  final double argument;
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: How do we decide whether to include this check?  In this case, I suspect it doesn't make it faster.

File: packages/flutter/lib/src/rendering/box.dart [hunk 2]
```diff
@@ -611,7 +703,25 @@ abstract class RenderBox extends RenderObject {
...
+  /// This function should only be called on one's children. Calling this
+  /// function couples the child with the parent so that when the child's layout
+  /// changes, the parent is notified (via [markNeedsLayout]).
+  ///
+  /// Calling this function is expensive and as it can result in O(N^2)
+  /// behavior.
+  ///
+  /// Do not override this function. Instead, implement
+  /// [computeMinIntrinsicHeight].
+  double getMinIntrinsicHeight(double width) {
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: Why not `mustCallSuper` like the width version?

File: packages/flutter/lib/src/rendering/box.dart [hunk 3]
```diff
@@ -623,36 +733,65 @@ abstract class RenderBox extends RenderObject {
...
+  /// This function should only be called on one's children. Calling this
+  /// function couples the child with the parent so that when the child's layout
+  /// changes, the parent is notified (via [markNeedsLayout]).
+  ///
+  /// Calling this function is expensive and as it can result in O(N^2)
+  /// behavior.
+  ///
+  /// Do not override this function. Instead, implement
+  /// [computeMaxIntrinsicHeight].
+  double getMaxIntrinsicHeight(double width) {
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/box.dart]: Why not `mustCallSuper` like the width version?

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/rendering/box.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: We should probably do an audit. I agree that this is a case where we shouldn't have it. It makes sense to have it when you're frequently reusing objects, or when the objects are really complex (ThemeData, for instance).

Filed https://github.com/flutter/flutter/issues/4447

File: packages/flutter/lib/src/rendering/box.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart]: fixed. this was a copypasta mistake.

**Pull request merged**